source/code/include/playfab/PlayFabEventBuffer.h

```diff
@@ -20,7 +20,7 @@ namespace PlayFab
             next(nullptr),
             eventIndex(index),
             timestamp(std::time(nullptr)), // current time
-            eventRequest(std::move(request))
```

Comment:
>std::move( [](start = 25, length = 10)

the std::move actually isn't redundant here. std::shared_ptr can be copied so without std::move here we end up doing a shared_ptr copies which unnecessarily manipulates the ref count.

the std::move would only be unnecessary if you made the parameter to the constructor an r-value reference but I think what you had before is probably simpler

Reply:
Since this is a shared_ptr, this std::move will ensure the original passed in request from the user would no longer be valid. Which brings up a question: Should we ever delete a user's shared_ptr? My only intention with shared_ptr is if we would share ownership, which kind of makes sense from the perspective of this object. (event request is also a shared_ptr, reinforcing this). 

BUT, this kind of ties into the const & idea. Herb Sutter did say passing by shared_ptr should only be by value if you care to transfer its ownership (This is what I meant earlier as making sense from this Object's perspective since we do want that ref count incremented by having this insternally as a shared_ptr). But const& if we don't really know if we will retain ownership, otherwise we should be passing the inner object's reference. 

So I think it is a real question as to why we should be passing this by a shared_ptr in the first place if ideally there's only ever one reference to that object (therefore this Packet should have contained a unique_ptr, which would make that std::move a more valid kind of thing). 

Sorry for the length, but this doesn't exactly end here though. This object is utilized by the PlayFabEventBuffer/Pipeline objects, which also prefer shared_ptr (which I think may increase the ref count around). 

Thoughts?

---
In reply to: [343929670](https://github.com/PlayFab/XPlatCoreTemplate/pull/41#discussion_r343929670) [](ancestors = 343929670)

Reply:
quick recap of offline chat we had:
* the rules for when to std::move vs not are a little soft. My personal preference is that the use of std::move or the absence of it for shared_ptrs should always indicate intent. std::move indicates "I am transferring ownership and no longer need to use this". a copy indicates "I am sharing ownership and still need to use this". In this way I think the std::move here should be retained. Herb Sutter's article did make a point about mentioning std::move-ing until you get to your destination. I think this operates under the assumption that the parameters that you are passing your shared_ptr through to its destination are r-value references (e.g. `std::shared_ptr<t>&&`). In which case that makes enough sense and the use of r-value references helps clarify when a callee is only intended to pass along a shared_ptr reference and not hold ownership of it.

* there was some discussion of using unique_ptr internally instead of std::shared_ptr. I am all onboard with using unique_ptr instead of shared_ptr because it means that there is somewhere a clear owner of the resource which makes that resource's lifetime easier to understand and way more predictable. I would be careful about doing it in such a way that isn't API breaking if possible. Changing from shared_ptr to unique_ptr may be a non-trivial task which requires some restructuring though since it may mean that some system which previously owned a resource may only have a weak reference to it once unique_ptr becomes your memory style.

* there was some discussion about the use of raw ptrs vs references. I think we are both onboard with using raw values to signal that a function does not have or manipulate ownership of a resource.

sounds like there is going to be a little bit of work before I continue to look at this PR so I will wait for the next iteration. 

---
In reply to: [343932433](https://github.com/PlayFab/XPlatCoreTemplate/pull/41#discussion_r343932433) [](ancestors = 343932433,343929670)